### PR TITLE
Added controller with zookeeper. Added DAL package

### PR DESCRIPTION
**Pull request by bristy**
**Changes in PR:**

1. Added controller election
Controller election is based on epoch (/controller_epoch in zk ). Controller data is stored at /controller node as ephemeral node. Which ever server updates the controller_epoch and /controller atomically becomes the controller.

2. Added DAL package.
3. Removed idea files.

### Test
Start zookeeper on local host
start server with server.properties file.

### CODE REVIEW

File: core/src/main/java/job/scheduler/core/JobScheduler.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Entry class for Job scheduler.
+ */
+@Slf4j
+public class JobScheduler {
+
+  private static ServerConfig getConfigFromProps() {
+    // TODO: read from props
+    return new ServerConfig(String.valueOf(1), "localhost:2181", 18000);
```
Comment (ankitsingh-wtg) [inline core/src/main/java/job/scheduler/core/JobScheduler.java]: Add yml config file and read it here. Instead of hardcoding.

Comment (kunal-exuberant) [reply core/src/main/java/job/scheduler/core/JobScheduler.java]: +1

File: core/src/main/java/job/scheduler/core/JobScheduler.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  public static void main(String[] args) {
+    try {
+      ServerConfig config = getConfigFromProps();
+      JobSchedulerServer server = new JobSchedulerServer(config);
+      server.startup();
+      server.awaitShutdown();
+    } catch (Throwable ex) {
+      log.error("Exiting server due to fatal exception.");
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/JobScheduler.java:59]: Add a constant file for all such literals

File: core/src/main/java/job/scheduler/core/JobScheduler.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      System.exit(1);
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/JobScheduler.java:60]: why is this needed?

File: core/src/main/java/job/scheduler/core/JobScheduler.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    System.exit(0);
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/JobScheduler.java:62]: why is this needed?

File: core/src/main/java/job/scheduler/core/config/ServerConfig.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package job.scheduler.core.config;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.NonNull;
+
+@Data
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/config/ServerConfig.java]: Only @Getter, please remove @Setter.
Make this class immutable.

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.List;
+
+/***
+ * Main controller class responsible to assign work to other servers.
+ */
+
+@Slf4j
+@RequiredArgsConstructor
+public class JobSchedulerController {
+  private static final String CONTROLLER_PARENT = "/controller";
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: Move the literals to constants file.

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  private static final String SERVER_PARENT = "/servers";
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: Move the literals to constants file.

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  private String getLeaderNodePath() throws KeeperException, InterruptedException {
+    List<String> guids = zkClient.getChildren(CONTROLLER_PARENT);
+    if (!guids.isEmpty()) {
+      String leaderGUID = Collections.min(guids);
+      log.info("leader is {}.", leaderGUID);
+      return CONTROLLER_PARENT + "/" + leaderGUID;
+    } else {
+      // should not happen
+      log.error("No server is registered.");
+      return null;
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: Return Optional instead of null

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+  private String getLeaderNodePath() throws KeeperException, InterruptedException {
+    List<String> guids = zkClient.getChildren(CONTROLLER_PARENT);
+    if (!guids.isEmpty()) {
+      String leaderGUID = Collections.min(guids);
+      log.info("leader is {}.", leaderGUID);
+      return CONTROLLER_PARENT + "/" + leaderGUID;
+    } else {
+      // should not happen
+      log.error("No server is registered.");
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: Move literal to constants

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      log.info("Current node {} is leader ", electedNodeId);
+      return electedNodeId;
+    }
+    List<String> guids = zkClient.getChildren(CONTROLLER_PARENT);
+    Collections.sort(guids);
+    int index = Collections.binarySearch(guids, electedNodeId);
+    if (index > 0) {
+      return guids.get(index - 1);
+    } else {
+      return null;
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: Avoid null return. Use optionals

File: core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  private class ControllerMonitor implements ZookeeperClient.ZHandler {
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/controller/JobSchedulerController.java]: What is the utility of private class?

File: core/src/main/java/job/scheduler/core/utils/TaskScheduler.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+  @Override
+  public boolean isStarted() {
+    synchronized(this) {
+      return executor != null;
+    }
+  }
+
+  @Override
+  public ScheduledFuture<?> schedule(String name, Runnable runnable, long delayMs, long periodMs, TimeUnit unit) {
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/utils/TaskScheduler.java:82]: add jobId

File: core/src/main/java/job/scheduler/core/utils/TaskScheduler.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    log.debug("Scheduling task %s with initial delay %d ms and period %d ms."
```
Comment (kunal-exuberant) [inline core/src/main/java/job/scheduler/core/utils/TaskScheduler.java:83]: mention job Id in comment too

**Review by kunal-exuberant — COMMENTED** [review]
Did one iteration

File: core/src/main/java/job/scheduler/core/JobScheduler.java (hunk 2 above)
Comment (bristy) [reply core/src/main/java/job/scheduler/core/JobScheduler.java:59]: These are error messages.

File: core/src/main/java/job/scheduler/core/JobScheduler.java (hunk 3 above)
Comment (bristy) [reply core/src/main/java/job/scheduler/core/JobScheduler.java:60]: To hard stop sever. Lifecycle is managed by us so if we want to exit with proper error/success. (Removing them will result same if we dont catch error.)

File: core/src/main/java/job/scheduler/core/utils/TaskScheduler.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+            .format(name, TimeUnit.MILLISECONDS.convert(delayMs, unit), TimeUnit.MILLISECONDS.convert(periodMs, unit)));
+    synchronized (this) {
+      ensureRunning();
+      if (periodMs >= 0) {
+        return executor.scheduleAtFixedRate(runnable, delayMs, periodMs, unit);
```
Comment (vivekbrnl) [inline core/src/main/java/job/scheduler/core/utils/TaskScheduler.java:88]: I guess, these schedule methods are already threadsafe
